Examples/NN/example_MLP_DLR_airfoil.py

```diff
@@ -0,0 +1,128 @@
+    '''
+    x, y = td[:]
+    pyLOM.pprint(0,f'name={name} ({len(td)}), x ({x.shape}) = [{x.min(dim=0)},{x.max(dim=0)}], y ({y.shape}) = [{y.min(dim=0)},{y.max(dim=0)}]')
+
+
+## Set device
+device = pyLOM.NN.select_device("cpu") # Force CPU for this example, if left in blank it will automatically select the device
+
+
+## Load datasets and set up the results output
+BASEDIR = './DATA'
+CASESTR = 'NRL7301'
+RESUDIR = 'MLP_DLR_airfoil'
+pyLOM.NN.create_results_folder(RESUDIR)
+
+scaler     = pyLOM.NN.MinMaxScaler()
+_,td_train = load_dataset(os.path.join(BASEDIR,f'{CASESTR}_TRAIN.h5'),scaler,scaler)
+_,td_test  = load_dataset(os.path.join(BASEDIR,f'{CASESTR}_TEST.h5'),scaler,scaler)
+_,td_val   = load_dataset(os.path.join(BASEDIR,f'{CASESTR}_VAL.h5'),scaler,scaler)
+
+print_dset_stats('train',td_train)
+print_dset_stats('test', td_test)
+print_dset_stats('val',  td_val)
+
+
+## define the optimizer
+optimizer = pyLOM.NN.OptunaOptimizer(
+    optimization_params={
+        "lr": (0.00001, 0.01),  # fixed parameter
+        "n_layers": (1, 4),  # optimizable parameter,
+        "batch_size": (128, 512),
+        "hidden_size": (200, 400),
+        "p_dropouts": (0.1, 0.5),
+        "epochs": 50,
+    },
+    n_trials=25,
+    direction="minimize",
+    pruner=optuna.pruners.MedianPruner(n_startup_trials=5, n_warmup_steps=5, interval_steps=1),
+    save_dir=None,
+)
+
+model = pyLOM.NN.MLP(
+    input_size=4,
+    output_size=1,
+    hidden_size=256,
+    n_layers=2,
+    p_dropouts=0.15,
+)
+
+training_params = {
+    "epochs": 150,
+    'lr_scheduler_step': 1,
+    "optimizer_class": torch.optim.Adam,
+    "loss_fn": torch.nn.MSELoss(),
+    "print_rate_epoch": 1,
+    "num_workers": 6,
+    "device": device,
+    "lr": 0.0008380427541690664, 
+    "lr_gamma": 0.9905178804615045, 
+    "batch_size": 119, 
+    "hidden_size": 129,
+    "n_layers": 6
+}
+
+pipeline = pyLOM.NN.Pipeline(
+    train_dataset=td_train,
+    test_dataset=td_test,
+    valid_dataset=td_val,
+    # To optimize the hyperparameters:
+    # optimizer=optimizer,
+    # model_class=pyLOM.NN.MLP,
+    # To train a model:
+    model=model,
+    training_params=training_params,
+)
+
+pipeline.run()
+
+
+## check saving and loading the model
+pipeline.model.save(os.path.join(RESUDIR,"model.pth"))
+model = pyLOM.NN.MLP.load(RESUDIR + "/model.pth")
+# to predict from a dataset
+preds = model.predict(td_test, batch_size=2048)
+# to predict from a tensor
+# preds = model(torch.tensor(dataset_test[:][0], device=model.device)).cpu().detach().numpy()
+scaled_preds = scaler.inverse_transform([preds])[0]
+scaled_y     = scaler.inverse_transform([td_test[:][1]])[0]
+
+# check that the scaling is correct
+print(scaled_y.min(), scaled_y.max())
+
+print(f"MAE: {np.abs(scaled_preds - np.array(scaled_y)).mean()}")
+print(f"MRE: {np.abs(scaled_preds - np.array(scaled_y)).mean() / abs(np.array(scaled_y).mean() + 1e-6)}")
+print(f"MSE: {((scaled_preds - np.array(scaled_y)) ** 2).mean()}")
```

Comment:
These could be added inside pyLOM.NN as utils since they can be helpful for other models

Reply:
it would be great, and even some kind of evaluation could be added to the pipeline to make it model independent. I think that here on UPM we are working on a more complex "statistical evaluation" that we can consider to add. If not, a simple set of metrics should be enough

Reply:
@DavidRamosArchilla let's add that as a `pyLOM.NN.stats` module for now

Reply:
we had this evaluator class that wraps a bunch of evaluation metrics and only depends on numpy
```python
class RegressionEvaluator(Evaluator):
    r"""
    Evaluator class for regression tasks. 
    Includes methods to calculate the mean squared error (MSE), mean absolute error (MAE), 
    mean relative error (MRE), quantiles of the absolute errors, L2 error, and R-squared.

    Args:
        tolerance (float): Tolerance level to consider values close to zero for MRE calculation (default: ``1e-4``).
    """

    def __init__(
        self,
        tolerance: float = 1e-4,
    ) -> None:
        self.tolerance = tolerance

    @property
    def tolerance(self) -> float:
        return self._tolerance
    
    @tolerance.setter
    def tolerance(self, value: float) -> None:
        if value < 0:
            raise ValueError("Tolerance must be a positive value.")
        if value == 0:
            raise ValueError("Tolerance cannot be zero.")
        if value > 1e-2:
            raise ValueError("Tolerance should be less than 1e-2.")
        self._tolerance = value

    def mean_squared_error(
        self,
        y_true: np.ndarray,
        y_pred: np.ndarray,
    ) -> float:
        """
        Compute the mean squared error (MSE) between the true values and the predicted values.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.

        Returns:
            float: The mean squared error.
        """
        return np.mean((y_true - y_pred) ** 2)
    
    def mean_absolute_error(self, y_true, y_pred):
        """
        Compute the mean absolute error (MAE) between the true values and the predicted values.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.

        Returns:
            float: The mean absolute error.
        """
        return np.mean(np.abs(y_true - y_pred))
    
    def mean_relative_error(
        self,
        y_pred: np.ndarray,
        y_true: np.ndarray,
    ) -> float:
        """
        Compute the mean relative error (MRE) between the true values and the predicted values,
        excluding cases where y_true is close to zero within a specified tolerance.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.
            tolerance (float): Tolerance level to consider values close to zero. Default is 1e-4.

        Returns:
            float: The mean relative error excluding cases where y_true is close to zero.
        """
        mask = np.abs(y_true) > self.tolerance
        y_true_masked = y_true[mask]
        y_pred_masked = y_pred[mask]
        
        relative_errors = np.abs((y_true_masked - y_pred_masked) / (y_true_masked))
        
        return np.mean(relative_errors) * 100
    
    def ae_q(
        self,
        y_pred: np.ndarray,
        y_true: np.ndarray,
        quantile: int,
    ) -> float:
        """
        Calculate the quantile of the absolute errors between the true and predicted values.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.
            quantile (int): The quantile to calculate. Must be between 0 and 100.

        Returns:
            float: The quantile of the absolute errors.
        """

        absolute_errors = np.abs(y_true - y_pred)
        return np.percentile(absolute_errors, quantile)    

    def l2_error(
        self,
        y_pred: np.ndarray,
        y_true: np.ndarray,
    ) -> float:
        """
        Calculate the L2 error between the true and predicted values.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.

        Returns:
            float: The L2 error.
        """

        return np.linalg.norm(y_true - y_pred) / np.linalg.norm(y_true)
    
    def R2(
        self,
        y_true: np.ndarray,
        y_pred: np.ndarray,
    ) -> float:
        """
        Calculate the R-squared (coefficient of determination) for a set of true and predicted values.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.

        Returns:
            float: The R-squared value.
        """
        y_mean = np.mean(y_true)
        total_sum_of_squares = np.sum((y_true - y_mean) ** 2)
        residual_sum_of_squares = np.sum((y_true - y_pred) ** 2)
        r_squared = 1 - (residual_sum_of_squares / total_sum_of_squares)
        return r_squared

    def print_metrics(self):
        """
        Print the calculated regression metrics.
        """
        if self._metrics is None:
            raise ValueError("No metrics have been calculated yet.")
        
        print("\nRegression evaluator metrics:")
        
        for key, value in self._metrics.items():
            if key == "mre":
                print(f"{key}: {value:.4f}%")
            elif key == "r2":
                print(f"{key}: {value:.4f}")
            else:
                if value < 1e-3 or value > 1e3:
                    print(f"{key}: {value:.4e}")
                else:
                    print(f"{key}: {value:.4f}")


    def __call__(
        self,
        y_true: np.ndarray,
        y_pred: np.ndarray,
        x: np.ndarray = None,
    ) -> dict:
        """
        Calculate multiple regression metrics between the true and predicted values.

        Args:
            y_true (numpy.ndarray): The true values.
            y_pred (numpy.ndarray): The predicted values.

        Returns:
            dict: A dictionary containing the calculated regression metrics.
        """
        try:
            y_true = np.array(y_true)
            y_pred = np.array(y_pred)
        except AttributeError:
            raise ValueError(f"could not create numpy arrays from object with type {type(y_true)}")
        
        mse = self.mean_squared_error(y_true, y_pred)
        rmse = np.sqrt(mse)
        mae = self.mean_absolute_error(y_true, y_pred)
        mre = self.mean_relative_error(y_true, y_pred)
        aq_95 = self.ae_q(y_true, y_pred, 95)
        aq_99 = self.ae_q(y_true, y_pred, 99)
        r2 = self.R2(y_true, y_pred)
        l2_error = self.l2_error(y_true, y_pred)
        self._metrics = {
            "mse": mse,
            "rmse": rmse,
            "mae": mae, 
            "mre": mre, 
            "ae_95": aq_95,
            "ae_99": aq_99,
            "r2": r2,
            "l2_error": l2_error 
        }
        return self._metrics
```
It can be easily extended with new metrics and it can be used in 2-3 lines of code. If you prefer something like this, i could add it

Reply:
Sounds good!